### PR TITLE
Adds plugin to add new cpts and taxonomies for the developer blog

### LINKED ISSUE #68: [WIP] Add CPT and Fields for Developer Hours posts
During [July meeting of the Editorial Group](https://make.wordpress.org/core/2024/07/05/developer-blog-editorial-meeting-summary-july-4-2024/) a [proposal to add Developer Hours recordings](https://github.com/WordPress/developer-blog-content/discussions/277) as posts to the Developer Blog was approved with the following implementation outline

- Create a new Custom Post Type with single post and archive page templates and add them to the normal post feed.
- Each Developer hours will get its individual page with YouTube video embed, shared resources posted and transcript added + Featured image, that is used for the Archive page.
- The Archive page lists each post with title, features image and date.
- As author we would use the main presenter's name.

In this issue and comments, I would like to kick off the discussion for how we approach the implementation.

### PR DESCRIPTION
**Pull request by ryanwelcher**
This PR adds a plugin to add new `dev-blog-videos` post type to the developer blog. 

Additionally it filters the main query to add this CPT to the main query and RSS feed

## Update

This PR now moves all of the cpts, taxonomies, and hooks into a single plugin as per @ryelle's suggestion.

Closes #68

### CODE REVIEW

File: source/wp-content/plugins/wporg-developer-blog-videos/wporg-developer-blog-videos.php
Comment (ndiego) [inline source/wp-content/plugins/wporg-developer-blog-videos/wporg-developer-blog-videos.php]:
```suggestion
			$query_vars['post_type'] = [ 'post', 'snippets', 'dev-blog-videos' ];
```

**Comment by justintadlock** [comment]
Should the CPT should have the `dev-blog-` prefix? We didn't do that for the snippets type. I'm not sure if it maybe had interfered with something else.

**Comment by ryanwelcher** [comment]
> Should the CPT should have the `dev-blog-` prefix? We didn't do that for the snippets type. I'm not sure if it maybe had interfered with something else.

I thought that `videos` was pretty generic. I'm not married to the name so happy to change it.

**Comment by ryelle** [comment]
> Should the CPT should have the dev-blog- prefix? We didn't do that for the snippets type. I'm not sure if it maybe had interfered with something else.

For "snippets", I figured the word was unique enough to not be a conflict (and if we wanted snippets on another site, we could just activate this plugin there, too). For "videos", that's more generic, and this implementation does seem more specific to the dev blog, so I would keep the prefix here.

**Review by justintadlock — APPROVED** [review]

**Comment by justintadlock** [comment]
Thanks for the feedback on the slug. Looks good to me.

File: source/wp-content/plugins/wporg-developer-blog-videos/wporg-developer-blog-videos.php
Comment (ryelle) [inline source/wp-content/plugins/wporg-developer-blog-videos/wporg-developer-blog-videos.php]: What about adding this CPT to the existing list? This way you don't need to worry about the order of these hooks. I would also suggest making the same change in the snippets CPT in another PR.

```suggestion
			$query_vars['post_type'][] = 'dev-blog-videos';
```

File: source/wp-content/plugins/wporg-developer-blog-videos/wporg-developer-blog-videos.php
Comment (ryelle) [inline source/wp-content/plugins/wporg-developer-blog-videos/wporg-developer-blog-videos.php]: Did you mean `is_front_page()`? I would also use the function version of `is_home`, for consistency (I had to look up whether that was correct).

File: source/wp-content/plugins/wporg-developer-blog-videos/wporg-developer-blog-videos.php
Comment (ryelle) [inline source/wp-content/plugins/wporg-developer-blog-videos/wporg-developer-blog-videos.php]: You could set this to `videos` for a better archive page URL, if you want to use that.

And maybe add a `rewrite` setting for a nicer permalink?

**Review by ryelle — COMMENTED** [review]
If you want to add this CPT into the homepage "Latest posts" section, you can do so with this hook:

```php
/*
 * Display videos in the query loop block on the homepage.
 */
add_action(
	'query_loop_block_query_vars',
	function( $query ) {
		if ( ! is_admin() && is_front_page() ) {
			$query['post_type'] = (array) $query['post_type'];
			$query['post_type'][] = 'dev-blog-videos';
		}
		return $query;
	}
);
```

I was also thinking about this update more, and I think you should refactor the `snippets` plugin to be a more general "dev blog CPTs" plugin instead of separating them out— especially if you might add even more CPTs. Once merged, these plugins are available on all wp.org sites, and it would just be neater if all the "deve blog" functionality was in one place. You're also repeating a few hooks here that might be more straightforward if they were only done once.

**Comment by ryanwelcher** [comment]
@ryelle thanks for the feedback on this and I agree that we should probably move all of this in to a revamped version of the snippets plugin. I'll do that today.

**Comment by ryanwelcher** [comment]
I've moved all of the changes into a single plugin. One thing to note was that I have all of the new CPTS showing up on the homepage loop. @justintadlock @bph @ndiego is that what we're looking for?

**Comment by bph** [comment]
> I have all of the new CPTS showing up on the homepage loop. Is that what we're looking for?

It's a great question. For now (AKA soft launch) I'd rather not have either of the CPTs on the homepage, so we can tweak things for a bit. 
I might have a preference for a more controlled design on how videos or snippets are displayed on the homepage instead of just reverse chronological. Once I see more snippets and videos posted, I might form a better opinion...

File: source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php
Comment (justintadlock) [inline source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php]: We'd also want to include `is_home()` to account for the blog/all-posts page: https://developer.wordpress.org/news/all-posts/

```php
if ( ! is_admin() && ( is_front_page() || is_home() ) ) {
```

File: source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php
Comment (justintadlock) [inline source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php]: Add `excerpt` support.

File: source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php
Comment (justintadlock) [inline source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php]: Add `excerpt` support.

**Comment by justintadlock** [comment]
I'm OK with holding off inclusion of posts on the front and homepages if @bph thinks it best for now.

I added a note to include excerpt support for the CPTs. It's something I overlooked previously.

Other than that, we probably need a decision on whether both or either of these should support comments.

**Comment by ryanwelcher** [comment]
> Other than that, we probably need a decision on whether both or either of these should support comments.

I think we should have comments for snippets but maybe not videos? The videos will be on YouTube so they can comment here. I'm not married to this idea though and we can always turn them off later.

**Comment by ryanwelcher** [comment]
I'm going to add commenting support for each CPT and we can go from there.

**Comment by ndiego** [comment]
> For now (AKA soft launch) I'd rather not have either of the CPTs on the homepage, so we can tweak things for a bit.

+1

**Review by justintadlock — APPROVED** [review]

**Review by ndiego — APPROVED** [review]
🚢

File: source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php
Comment (ryelle) [inline source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php]: This can be slightly shorthanded, since anyone seeing this should know "CPT", and the description has the details.

```suggestion
 * Plugin Name: Developer Blog CPTs
```

File: source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php
Comment (ryelle) [inline source/wp-content/plugins/wporg-developer-blog-cpts/wporg-developer-blog-cpts.php]: Minor change since this function has been updated:

```suggestion
 * Register the post types and taxonomies.
```

**Review by ryelle — COMMENTED** [review]
Looks good, thanks for combining the two plugins. Two super-minor notes, if you approve those I can deploy this in the morning 🙂

**Comment by ryanwelcher** [comment]
@ryelle thanks for the review and suggestions!

**Review by ryelle — APPROVED** [review]

**Comment by ryelle** [comment]
This has been deployed and activated on the dev blog 👍🏻

**Pull request merged**
